src/page.c

```diff
@@ -737,9 +737,15 @@ static _Atomic(void*) deferred_arg; // = NULL
 
 void _mi_deferred_free(mi_heap_t* heap, bool force) {
   heap->tld->heartbeat++;
-  if (deferred_free != NULL && !heap->tld->recurse) {
+  if (!heap->tld->recurse) {
     heap->tld->recurse = true;
-    deferred_free(force, heap->tld->heartbeat, mi_atomic_load_ptr_relaxed(void,&deferred_arg));
+    if (deferred_free != NULL) {
+      deferred_free(force, heap->tld->heartbeat, mi_atomic_load_ptr_relaxed(void,&deferred_arg));
+    }
+    if (heap->deferred_free != NULL) {
+      // TODO: Should heap->deferred_free_arg be an atomic load? 
+      (heap->deferred_free)(heap, force, heap->tld->heartbeat, heap->deferred_free_arg);
```

Comment:
I am not sure if it causes confusions since there are two function invocations with the same name. i.e., 
* `deferred_free(force, heap->tld->heartbeat, mi_atomic_load_ptr_relaxed(void,&deferred_arg))`
* `(heap->deferred_free)(heap, force, heap->tld->heartbeat, heap->deferred_free_arg)`

Reply:
I don't think it causes confusion? First call is call to global variable that holds function pointer and second call is call to other function pointer that is stored inside heap instance.

Reply:
I added documentation to invocation of `deferred_free` and `heap->deferred_free`